### PR TITLE
Handling Azure credentials in a secure way

### PR DESCRIPTION
**Pull request by eduardoj**
Azure credentials will be stored encrypted (with OpenSSL library) in the database.

This was a work done by @mschnitzer.

This pull request superseeds #4603.

### CODE REVIEW

**Comment by eduardoj** [comment]
[P2: [3] P6: (5) Handle Azure credentials in a secure way](https://trello.com/c/FfTd7Dhu/666-p2-3-p6-5-handle-azure-credentials-in-a-secure-way)

File: src/api/config/routes.rb [hunk 1]
```diff
@@ -210,6 +210,9 @@ def self.public_or_about_path?(request)
 
         resource :log, only: :show, controller: 'webui/cloud/upload_job/logs'
       end
+      resource :azure do
```
Comment (ChrisBr) [inline src/api/config/routes.rb]: IIRC if you don't use the `` only: []`` it will create routes for azure as well.

File: src/api/spec/models/cloud/azure/configuration_spec.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+# WARNING: Some tests require real backend answers, so make sure you uncomment
+# this line and start a test backend.
+# CONFIG['global_write_through'] = true
+
+RSpec.describe Cloud::Azure::Configuration, type: :model, vcr: true do
+  describe 'data encryption' do
+    let(:config) { create(:azure_configuration, application_id: 'Hey OBS!', application_key: 'Hey OBS?') }
+    let(:secret_key) { OpenSSL::PKey::RSA.new(File.read(Rails.root.join('spec', 'support', 'files', 'cloudupload_secret_key_tests.pem'))) }
```
Comment (ChrisBr) [inline src/api/spec/models/cloud/azure/configuration_spec.rb]: I don't understand why the model uses this key? I would expect that we need to mock a backend call?
Also I don't like that we create always a new key, why is this necessary as we already commit the key to git?

File: src/api/app/models/cloud/azure/configuration.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+      before_save :encrypt_credentials
+
+      belongs_to :user
+
+      def self.table_name_prefix
+        'cloud_azure_'
+      end
+
+      def available?
```
Comment (ChrisBr) [inline src/api/app/models/cloud/azure/configuration.rb]: Are you sure this works? As we have presence validations this should be always true (after the user visited the show page).

Comment (eduardoj) [reply src/api/app/models/cloud/azure/configuration.rb]: @ChrisBr 

> Are you sure this works? As we have presence validations this should be always true (after the user visited the show page).

Tested again. `available?` is used when the user does not have any credential stored. You can see it in the images attached:

azure_configuration_1

azure_configuration_2

File: src/api/config/routes.rb (hunk 1 above)
Comment (eduardoj) [reply src/api/config/routes.rb]: `only: []` added.

File: src/api/spec/models/cloud/azure/configuration_spec.rb (hunk 2 above)
Comment (eduardoj) [reply src/api/spec/models/cloud/azure/configuration_spec.rb]: @ChrisBr 

> I don't understand why the model uses this key? I would expect that we need to mock a backend call?

Done. The request done by the backend call is now stubbed.

> Also I don't like that we create always a new key, why is this necessary as we already commit the key to git?

The secret_key is in a let clause. And now the public_key is also in a let clause. So reading the keys will be executed one time.

File: src/api/config/routes.rb [hunk 4]
```diff
@@ -210,6 +210,9 @@ def self.public_or_about_path?(request)
 
         resource :log, only: :show, controller: 'webui/cloud/upload_job/logs'
       end
+      resource :azure, only: [] do
```
Comment (hennevogel) [inline src/api/config/routes.rb]: :thought_balloon: [scope](http://guides.rubyonrails.org/routing.html#prefixing-the-named-route-helpers)?

Comment (eduardoj) [reply src/api/config/routes.rb]: @hennevogel 

> :thought_balloon: scope?

Done.

Comment (eduardoj) [reply src/api/config/routes.rb]: @hennevogel 

> :thought_balloon: scope?

Moved to #4654.

**Comment by eduardoj** [comment]
Codecov hung up. Merging.
